src/testing/vortex/supervisor.zig

```diff
@@ -0,0 +1,336 @@
+//! If you have permissions troubles with Ubuntu, see:
+//! https://github.com/YoYoGames/GameMaker-Bugs/issues/6015#issuecomment-2135552784
+//!
+//! Further possible work:
+//!
+//! * full partitioning
+//! * filesystem faults
+//! * clock faults
+//! * upgrades (replicas and clients)
+//! * liveness checks
+//! * multiple drivers? could use a special multiplexer driver that delegates to others
+
+const std = @import("std");
+const builtin = @import("builtin");
+const LoggedProcess = @import("./logged_process.zig");
+const Replica = @import("./replica.zig");
+const NetworkFaults = @import("./network_faults.zig");
+const arbitrary = @import("./arbitrary.zig");
+
+const log = std.log.scoped(.supervisor);
+
+const process = std.process;
+const assert = std.debug.assert;
+
+const cluster_id = 1;
+const replica_count = 3;
+const replica_ports = [replica_count]u16{ 3000, 3001, 3002 };
+const replica_addresses = comma_separate_ports(&replica_ports);
+
+pub const CLIArgs = struct {
+    tigerbeetle_executable: []const u8,
+    test_duration_minutes: u16 = 10,
+};
+
+pub fn main(allocator: std.mem.Allocator, args: CLIArgs) !void {
+    if (builtin.os.tag == .windows) {
+        log.err("vortex is not supported for Windows", .{});
+        return error.NotSupported;
+    }
+
+    const tmp_dir = try create_tmp_dir(allocator);
+    defer allocator.free(tmp_dir);
+    defer std.fs.cwd().deleteTree(tmp_dir) catch |err| {
+        log.err("failed deleting temporary directory ({s}): {any}", .{ tmp_dir, err });
+    };
+
+    // Check that we are running as root.
+    const user_id = std.os.linux.getuid();
+    if (user_id != 0) {
+        log.err(
+            "this script needs to be run in a separate namespace using 'unshare -nfr'",
+            .{},
+        );
+        process.exit(1);
+    }
+
+    try NetworkFaults.setup(allocator);
+
+    log.info(
+        "starting test with target runtime of {d}m",
+        .{args.test_duration_minutes},
+    );
+    const test_duration_ns = @as(u64, @intCast(args.test_duration_minutes)) * std.time.ns_per_min;
+    const test_deadline = std.time.nanoTimestamp() + test_duration_ns;
+
+    var replicas: [replica_count]*Replica = undefined;
+    inline for (0..replica_count) |i| {
+        var datafile_buf: [1024]u8 = undefined;
+        const datafile = try std.fmt.bufPrint(
+            datafile_buf[0..],
+            "{s}/{d}_{d}.tigerbeetle",
+            .{ tmp_dir, cluster_id, i },
+        );
+
+        // Format each replica's datafile.
+        const result = try process.Child.run(.{ .allocator = allocator, .argv = &.{
+            args.tigerbeetle_executable,
+            "format",
+            std.fmt.comptimePrint("--cluster={d}", .{cluster_id}),
+            std.fmt.comptimePrint("--replica={d}", .{i}),
+            std.fmt.comptimePrint("--replica-count={d}", .{replica_count}),
+            datafile,
+        } });
+        defer allocator.free(result.stderr);
+        defer allocator.free(result.stdout);
```

Comment:
Same here, merge defers. Defer can not fail, so it is always correct to merge them.